### PR TITLE
Big Spellbook

### PR DESCRIPTION
**Pull request by Laserlicht**
HD thing.

Big spellbook is created on the fly. Not perfect, but we don't need additional assets.

Can disabled via `enableUiEnhancements`-Flag.

image
image

### CODE REVIEW

File: client/windows/CSpellWindow.cpp [hunk 1]
```diff
@@ -182,42 +200,46 @@ CSpellWindow::CSpellWindow(const CGHeroInstance * _myHero, CPlayerInterface * _m
...
+	interactiveAreas.push_back(std::make_shared<InteractiveArea>( Rect( 418 + pos.x + (isBigSpellbook ? 142 : 0), 405 + pos.y + offB, isBigSpellbook ? 60 : 36, 56), std::bind(&CSpellWindow::fmanaPtsb,      this),    459, this));
+	interactiveAreas.push_back(std::make_shared<InteractiveArea>( Rect( 549 + pos.x + offR,  94 + pos.y, 36, 56), std::bind(&CSpellWindow::selectSchool,   this, 0), 454, this));
+	interactiveAreas.push_back(std::make_shared<InteractiveArea>( Rect( 549 + pos.x + offR, 151 + pos.y, 45, 35), std::bind(&CSpellWindow::selectSchool,   this, 3), 457, this));
+	interactiveAreas.push_back(std::make_shared<InteractiveArea>( Rect( 549 + pos.x + offR, 210 + pos.y, 45, 35), std::bind(&CSpellWindow::selectSchool,   this, 1), 455, this));
+	interactiveAreas.push_back(std::make_shared<InteractiveArea>( Rect( 549 + pos.x + offR, 270 + pos.y, 45, 35), std::bind(&CSpellWindow::selectSchool,   this, 2), 456, this));
+	interactiveAreas.push_back(std::make_shared<InteractiveArea>( Rect( 549 + pos.x + offR, 330 + pos.y, 45, 35), std::bind(&CSpellWindow::selectSchool,   this, 4), 458, this));
```
Comment (IvanSavenko) [inline client/windows/CSpellWindow.cpp:218]: Perhaps create Rect separately and add offset to it in these calls, e.g:
```cpp
Rect schoolRect( 549 + pos.x + offR, 94 + pos.y, 36, 56);
interactiveAreas.push_back( schoolRect + Point(0, 0);
interactiveAreas.push_back( schoolRect + Point(0, 57);
...
```

**Review by IvanSavenko — COMMENTED** [review]
- No new assets required
- Available by default without forcing player to install mods

Many thanks for this approach :)

File: client/widgets/TextControls.cpp [hunk 2]
```diff
@@ -461,6 +461,9 @@ CGStatusBar::CGStatusBar(int x, int y, const ImagePath & name, int maxw)
 	addUsedEvents(LCLICK);
 
 	OBJECT_CONSTRUCTION_CAPTURING(255 - DISPOSE);
+
+	if(name.empty())
+		return;
```
[the remark this answers is not in the document]
Comment (Laserlicht) [reply client/widgets/TextControls.cpp]: Made name as optional.

[the remark this answers is not in the document]
Comment (Laserlicht) [reply client/widgets/TextControls.cpp]: It works good without background in spell window.

[the remark this answers is not in the document]
Comment (IvanSavenko) [reply client/widgets/TextControls.cpp]: Still should probably be a separate constructor.
Right now you have method that acts very differently depending on input parameters (filename)

[the remark this answers is not in the document]
Comment (Laserlicht) [reply client/widgets/TextControls.cpp]: Done :)

**Review by IvanSavenko — APPROVED** [review]

**Pull request merged**